File: wiss/test/old/mod/test3.m


module main;

use
#include "wiss_mc.d"

#include "wiss_mc.h"

(* wiss file name *)
#define FILENAME	adr("student")
(* wiss device name *)
#define DEVICENAME	adr("wiss.disk")
(* error handler *)
#define	checkerror(p,c)	if (c < 0) then printf("error %d in %s\n",c,p); halt;end

const
	SIDOFFSET	= 0;
	SAGEOFFSET	= 4;
	SYROFFSET	= 8;
	SNAMEOFFSET	= 12;
	SSEXOFFSET	= 60;

	AGEINDEX	= 0;	(* number for the index on age *)

(* student record format (for the wiss file) *)
type
	studentrec = record
		id : integer;	
		age : integer;		
		yr  : integer;
		name : array 0 : 47 of char;
		sex : char;
	end (* studentrec *);

(* variables used by the main program *)
var
	errorcode : integer;		(* for returned errors *)
	vol : integer;			(* volume id *)
	dummy : integer;
	condition : array 0 : 1 of BOOLEXP;
	keyattr : FIELDDESC;
	lb : DATADESC;


(* populate a sample relation *)
procedure load_database(vol : integer; filename : addr);

(* 
   [Description of the data file] 
    the test data is in an ascii file with the following format :
    each record per line; 
    record length : 34,
    0-2 : ID, 3-22 : name; 23 : sex; 25-26 : age, 28 : year, 
    29- : ignored for the time being
*)

const
	INPUT	= "/usr4/katz/wiss/test/2/btree/studentdata/student";
	SRECLENGTH	= 34;	/* length of a record */

var
	i : integer;			(* loop index *)
	errorcode : integer;		(* for returned error codes *)
	unixfile : integer;		(* unix file descriptor *)
	openfile_number : integer;	(* open file number of the wiss file *)
	buf : array 0 : SRECLENGTH of char;	
					(* buffer for records in unix file *)
	newrid : RID;			(* RID of the newly created record *)
	student : studentrec;		(* buffer for records in wiss file  *)

begin

	(* open the wiss file with WRITE permission *)
	openfile_number := wiss_openfile(vol, filename, WRITE);
	checkerror("test3/wiss_openfile", openfile_number);

	(* open the unix file first *)
	unixfile := open(INPUT, 0);
	if (unixfile < 0) then
		printf("\ncan not open student data file");
		halt;
	end;

	(* read records from the unix file and put them into the wiss file *)
	while (read(unixfile, buf, SRECLENGTH) > 0) do
		(* convert the record format *)
		i := 0;
		student.id := 0;
		while (i < 3) do
			if (buf[i] <> ' ') then 
				student.id := student.id * 10 + 
					(integer(buf[i]) - integer('0'));
			end; (* if *)
			inc(i);
		end (* while *);
			
		while (i < 23) do
			student.name[i - 3] := buf[i];
			inc(i);
		end (* while *);

		student.sex := buf[23];
		student.age := (integer(buf[25]) - integer('0')) * 10 +
				(integer(buf[26]) - integer('0'));
		student.yr  := (integer(buf[28]) - integer('0'));

		(* insert it into the wiss file *)
		errorcode := wiss_insertrecord(openfile_number, 
			adr(student), size(student), NULL, newrid);
		checkerror("test3/wiss_insertscan", errorcode);

	end; (* while *)

	(* close the wiss file *)
	errorcode := wiss_closefile(openfile_number);
	checkerror("test3/wiss_closefile", errorcode);

end load_database;



(* this procedure does a sequential scan over the relation call filename *)
procedure retrieve(vol : integer; filename : addr; boolexp : addr;
	keyattr, lb, ub: addr);
var
	errorcode : integer;		(* for returned error codes *)
	openfile_number : integer;	(* open file number of the wiss file *)
	indexfile_number : integer;	(* open file number of the index file *)
	scanid : integer;		(* id of the scan opened *)
	student : studentrec;		(* buffer for student records *)
	dummy : RID;
begin

	(* open the wiss file with READ permission *)
	openfile_number := wiss_openfile(vol, filename, READ);
	checkerror("test3/wiss_openfile", openfile_number);

	(* open the index file too *)
	indexfile_number := wiss_openindex(vol, filename, AGEINDEX, READ);
	checkerror("test3/wiss_openindex", indexfile_number);

	(* open an index scan on the file just opened *)
	scanid := wiss_openindexscan(openfile_number, indexfile_number, 
		keyattr, lb, ub, boolexp);
	checkerror("test3/wiss_openfilescan", scanid);

	(* set the scan cursor to the first qualified record if any *)
	errorcode := wiss_fetchfirst(scanid, dummy);

	(* do a sequential scan, and print out all the records *)
	while (errorcode >= eNOERROR) do

		(* read in the qualified record *)
		errorcode := wiss_readscan(scanid, adr(student), size(student));
		checkerror("test3/wiss_readscan", errorcode);

		(* print student record *)
		printf(" (ID = %3d, Age = %2.2d, Year = %d, Sex = %c, ",
			student.id, student.age, student.yr, student.sex);
		printf("Name = ""%-20.20s"")\n", student.name);

		(* set cursor to the next record *)
		errorcode := wiss_fetchnext(scanid, dummy);

	end (* while *);

	(* close the scan *)
	errorcode := wiss_closescan(scanid);
	checkerror("test3/wiss_closescan", errorcode);

	errorcode := wiss_closefile(openfile_number);
	checkerror("test3/wiss_closefile", errorcode);

	errorcode := wiss_closefile(indexfile_number);
	checkerror("test3/wiss_closefile", errorcode);

end retrieve;


(* start of the main module body *)
begin
	(* warm up procedure for wiss *)
	wiss_init;

	(* mount the volume call DEVICENAME *)
	vol := wiss_mount(DEVICENAME);	
	checkerror("test3/wiss_mount", vol);

	(* create a wiss file which has 6 pages initially, both the extent
	    fill factor and page fill factor are 90%                    *)
	errorcode := wiss_createfile(vol, FILENAME, 6, 90,90);
	checkerror("test3/wiss_createfile", errorcode);

	(* load the database from a unix file, and put it into a wiss file
	   call "FILENAME"						*)
	printf(" loading the database ... \n");
	load_database(vol, FILENAME);

	(* create an index on age *)
	assign2(keyattr.offset, SAGEOFFSET);
	assign2(keyattr.length, 4);
	keyattr.ftype := TINTEGER;
	errorcode := wiss_createindex(vol, FILENAME, AGEINDEX,
		adr(keyattr), 100 (* leaves 100 % full *), 
		FALSE (* not primary *), FALSE (* don't sort the file *));
	checkerror("test3/wiss_createindex", errorcode);

	(* build a boolean expression: find records of the female students 
	   who's age are greater than 25                                 *)

	condition[0].operator := EQ;
	assign2(condition[0].fielddesc.offset, SSEXOFFSET);
	assign2(condition[0].fielddesc.length, 1);
	condition[0].fielddesc.ftype := TSTRING;
	condition[0].bvalue[0] := 'f';
	condition[0].next := NULL;

	lb.dtype := TINTEGER;
	assign2(lb.length, 4);
	dummy := 25;
	movebytes(adr(lb.dvalue), adr(dummy), 4);

	(* RETRIEVE FROM (FILENAME) WHERE <condition> *)
	printf(" retrieving records of female students with age > 25 ... \n");
	retrieve(vol, FILENAME, adr(condition), adr(keyattr), adr(lb), NULL);

	(* remove the files since the test in done *)
	errorcode := wiss_destroyfile(vol, FILENAME);
	checkerror("test3/wiss_destroyfile", errorcode);
	errorcode := wiss_dropindex(vol, FILENAME, AGEINDEX);
	checkerror("test3/wiss_dropindex", errorcode);

	(* dismount the volume call DEVICENAME *)
	errorcode := wiss_dismount(DEVICENAME);  
	checkerror("test3/wiss_dismount", errorcode);


end main.

File: wiss/test/old/2/btree/Test3.c
#include <wiss.h>
#include <st.h>	

#define	SIDOFFSET	0
#define	SNAMEOFFSET	SIDOFFSET + sizeof(int)
#define	SSEXOFFSET	SNAMEOFFSET + 50
#define	SAGEOFFSET	SSEXOFFSET + 2
#define	SYROFFSET	SAGEOFFSET + sizeof(int)
#define SGPAOFFSET	SYROFFSET + sizeof(int)
#define	SRECLENGTH	34

/* error handlers */
#define	IOERROR(p,c)	if((int)(c)<0) \
		{printf("%s %s\n", p, io_error(c));io_final(); exit(-1);}
#define	BFERROR(p,c)	if((int)(c)<0) \
	{printf("%s %s\n",p,bf_error(c));bf_final(); io_final(); exit(-1);}
#define	ERROR(p,c)	if((int)(c)<0) printf("%s %s\n", p, st_error((int)(c)))
#define	STERROR(p,c)	if((int)(c)<0) {ERROR(p, c); st_final(); bf_final();\
				io_final();}


/* wiss file name */
#define	TESTFILE	"student"

/* print student record */
#define	PRINTST(student)\
	printf("(ID=%3d,Age=%2.2d,Year=%d,Sex=%c,",\
		student.id, student.age, student.yr,\
		student.sex[0] == 'm' ? 'M' : 'F');\
	printf("GPA=%4.2f,Name=\"%-20.20s\")\n",student.gpa, student.name)
	

/* student record format (for the wiss file) */
typedef	struct
	{
		int 	id;	
		char 	name[50];
		char 	sex[2];
		int 	age;		
		int 	yr;
		float 	gpa;
	} studentrec;

static	KEYINFO	keyattr[] ={ 	{ SIDOFFSET, sizeof(int), TINTEGER},
				{ SNAMEOFFSET, 30, TSTRING},
				{ SGPAOFFSET, sizeof(float), TFLOAT},
			   };


/************************************************************************
**
**				test3.c
** Description:
*************************************************************************/

main(argc, argv)
int	argc;	
char	**argv;
{
	int	errorcode;		/* for returned errors */
	int	vol;			/* volume id */
	int	i;
	int	oldflag;

	/* warm up procedure for wiss */
	wiss_checkflags(&argc,&argv);
	i = io_init();			/* initialize level 0 */
	IOERROR("test3/io_init", i);
	
	i = bf_init();		/* initialize level 1 */
	BFERROR("test3/bf_init", i);

	i = st_init();			/* initialize level 2 */
	STERROR("test3/st_init", i);


	/* mount the volume call "wiss2" */
	vol = st_mount("wiss2");	
	STERROR("test3/st_mount", vol);

/* step 1. create an empty data file */
	printf(" creating an empty data file %s\n", TESTFILE);
	errorcode = st_createfile(vol, TESTFILE, 6, 90,90);
	STERROR("test3/st_createfile", errorcode);


/* step 2. create 3 (empty) index files */
	printf(" creating three indices\n");
	/* index on id */
	errorcode = st_createindex(vol, TESTFILE, 1, &keyattr[0], 
			50, FALSE, FALSE);
	STERROR("test3/st_createindex", errorcode);

	/* index on name */
	errorcode = st_createindex(vol, TESTFILE, 2, &keyattr[1], 
			50, FALSE, FALSE);
	STERROR("test3/st_createindex", errorcode);

	/* index on GPA */
	errorcode = st_createindex(vol, TESTFILE, 3, &keyattr[2], 
			50, FALSE, FALSE);
	STERROR("test3/st_createindex", errorcode);



/* step 3. load the data base -- fill the data file */
	printf(" loading the database ... \n");
	load_database(vol, TESTFILE);

	/* RETRIEVE FROM (TESTFILE) */
	printf(" retrieving all records ... \n");
	retrieve(vol, TESTFILE, NULL);



/* step 4. create 3 index files by insertion */
	printf(" creating 3 index file by insertion \n");
	insertindex(vol, TESTFILE, 1);
	insertindex(vol, TESTFILE, 2);
	insertindex(vol, TESTFILE, 3);

/* step 5. delete all indices and drop the index files */

	printf(" deleting all the index files\n");
	deleteindex(vol, TESTFILE, 1);
	deleteindex(vol, TESTFILE, 2);
	deleteindex(vol, TESTFILE, 3);

	errorcode = st_dropbtree(vol, TESTFILE, 1);
	STERROR("test3/st_dropbtree", errorcode);
	errorcode = st_dropbtree(vol, TESTFILE, 2);
	STERROR("test3/st_dropbtree", errorcode);
	errorcode = st_dropbtree(vol, TESTFILE, 3);
	STERROR("test3/st_dropbtree", errorcode);

/* step 6. create 3 index files directly (from bottom up) */

	printf(" creating three index files in batch\n");
	errorcode = st_createindex(vol, TESTFILE, 1, 
			&keyattr[0], 50, FALSE, FALSE);
	STERROR("test3/st_createindex", errorcode);
	errorcode = st_createindex(vol, TESTFILE, 2, 
			&keyattr[1], 50, FALSE, FALSE);
	STERROR("test3/st_createindex", errorcode);
	errorcode = st_createindex(vol, TESTFILE, 3, 
			&keyattr[2], 50, FALSE, FALSE);
	STERROR("test3/st_createindex", errorcode);


/* step 7. delete all indices and drop the index files */

	printf(" deleting all the index files\n");
	deleteindex(vol, TESTFILE, 1);
	deleteindex(vol, TESTFILE, 2);
	deleteindex(vol, TESTFILE, 3);

	errorcode = st_dropbtree(vol, TESTFILE, 1);
	STERROR("test3/st_dropbtree", errorcode);
	errorcode = st_dropbtree(vol, TESTFILE, 2);
	STERROR("test3/st_dropbtree", errorcode);
	errorcode = st_dropbtree(vol, TESTFILE, 3);
	STERROR("test3/st_dropbtree", errorcode);


/* clear up */
	/* remove the file since the test in done */
	errorcode = st_destroyfile(vol, TESTFILE);
	STERROR("test3/st_destroyfile", errorcode);

	/* dismount the volume call "wiss2" */
	errorcode = st_dismount("wiss2");  
	STERROR("test3/st_dismount", errorcode);

}

retrieve(vol, filename)
int 	vol;		/* volume id */
char	*filename;	/* where the records are to be retrieved */
{
	int	errorcode;		/* for returned error codes */
	int 	openfile_number;	/* open file number of the wiss file */
	RID	currid;
	studentrec student;		/* buffer for student records */

	/* open the wiss file with READ permission */
	openfile_number = st_openfile(vol, filename, READ);
	STERROR("test3/st_openfile", openfile_number);


	/* do a sequential scan, and print out those qualified records */
	for ( errorcode = st_firstfile(openfile_number, &currid); 
					errorcode >= eNOERROR;)
	{
		/* read in the qualified record */
		errorcode = st_readrecord(openfile_number, &currid,
				(char *)&student, sizeof(studentrec));
		STERROR("test3/st_readrecord", errorcode);

		/* print student record */
		PRINTST(student);

		/* set cursor to the next qualified record */
		errorcode = st_nextfile(openfile_number, &currid, &currid);

	}
	printf("\n");

	errorcode = st_closefile(openfile_number);
	STERROR("test3/st_closefile", errorcode);

}

load_database(vol, filename)
int	vol;		/* volume id */
char	*filename;	/* wiss file to store the database */
{

#define	INPUT	"studentdata/student"

	int	i;			/* loop index */
	int	errorcode;		/* for returned error codes */
	int	unixfile;		/* unix file id */
	int	openfile_number;	/* open file number of the wiss file */
	char	buf[100];		/* buffer for records in unix file */
	RID	*newrid;		/* RID of the newly created record */
	studentrec	student;	/* buffer for records in wiss file  */

	/* open the wiss file with WRITE permission */
	openfile_number = st_openfile(vol, filename, WRITE);
	STERROR("test3/st_openfile", openfile_number);

	/* open the unix file first */
	unixfile = open(INPUT, 0);
	if (unixfile < 0)
	{
		printf("\ncan not open student data file");
		return;
	}

	/* read records from the unix file and put them into the wiss file */
	while (read(unixfile, buf, SRECLENGTH))
	{
		/* convert the record format */
		sscanf(buf,"%d", &student.id);
		for (i = 0; i < 20; i++)
			student.name[i] = buf[3 + i]; 		
		sscanf(&buf[23], "%c %2d %d %f", 
			 student.sex, &student.age, &student.yr, &student.gpa);

		/* insert it into the wiss file */
		errorcode = st_insertrecord(openfile_number, 
			&student, sizeof(studentrec), NULL, &newrid);
		STERROR("test3/st_insertrecord", errorcode);
	}

	/* close the wiss file */
	errorcode = st_closefile(openfile_number);
	STERROR("test3/st_closefile", errorcode);

}

/* create the whole index file by insertion */
insertindex(vol, filename, index)
int 	vol;		/* volume id */
char	*filename;	/* where the records are to be retrieved */
int	index;		/* index number */
{
	int	errorcode;		/* for returned error codes */
	int 	openfile_number;	/* open file number of the wiss file */
	int	indexfile_number;
	RID	currid;
	studentrec student;		/* buffer for student records */
	PID	pid;
	FID	fid;
	KEY	key;

	/* open the wiss file with READ permission */
	openfile_number = st_openfile(vol, filename, READ);
	STERROR("test3/st_openfile", openfile_number);

	indexfile_number = st_openbtree(vol, filename, index, WRITE);
	STERROR("test3/st_openbtree", indexfile_number);


	/* do a sequential scan, and insert all the indices */
	for ( errorcode = st_firstfile(openfile_number, &currid); 
					errorcode >= eNOERROR;)
	{
		/* read in the qualified record */
		errorcode = st_readrecord(openfile_number, &currid,
				(char *)&student, sizeof(studentrec));
		STERROR("test3/st_readrecord", errorcode);

		key.length = keyattr[index-1].length;
		key.type = keyattr[index-1].type;
		movebytes(key.value, 
			((char *)&student)+keyattr[index-1].offset, key.length);
		
		errorcode = st_insertindex(indexfile_number, &key, &currid);
		STERROR("test3/st_insertindex", errorcode);

		/* set cursor to the next qualified record */
		errorcode = st_nextfile(openfile_number, &currid, &currid);

	}
	printf("\n");

	/* close the index file */
	errorcode = st_closefile(indexfile_number);
	STERROR("test3/st_closefile", errorcode);
	errorcode = st_closefile(openfile_number);
	STERROR("test3/st_closefile", errorcode);

}



/* delete all the entries in the index file one by one */
deleteindex(vol, filename, index)
int 	vol;		/* volume id */
char	*filename;	/* where the records are to be retrieved */
int	index;		/* index # of the index file */
{
	int	errorcode;		/* for returned error codes */
	int 	openfile_number;	/* open file number of the wiss file */
	int	indexfile_number;
	RID	currid;
	studentrec student;		/* buffer for student records */
	PID	pid;
	FID	fid;
	KEY	key;


	/* open the wiss file with READ permission */
	openfile_number = st_openfile(vol, filename, READ);
	STERROR("test3/st_openfile", openfile_number);

	indexfile_number = st_openbtree(vol, filename, index, WRITE);
	STERROR("test3/st_openbtree", indexfile_number);


	/* do a sequential scan, and print out those qualified records */
	for ( errorcode = st_firstfile(openfile_number, &currid); 
					errorcode >= eNOERROR;)
	{
		/* read in the qualified record */
		errorcode = st_readrecord(openfile_number, &currid,
				(char *)&student, sizeof(studentrec));
		STERROR("test3/st_readrecord", errorcode);

		key.length = keyattr[index-1].length;
		key.type = keyattr[index-1].type;
		movebytes(key.value, 
			((char *)&student)+keyattr[index-1].offset, key.length);
		errorcode = st_deleteindex(indexfile_number, &key, &currid);
		STERROR("test3/st_deleteindex", errorcode);

		/* set cursor to the next qualified record */
		errorcode = st_nextfile(openfile_number, &currid, &currid);

	}
	printf("\n");

	/* close the index file */
	errorcode = st_closefile(indexfile_number);
	STERROR("test3/st_closefile", errorcode);
	errorcode = st_closefile(openfile_number);
	STERROR("test3/st_closefile", errorcode);

}
